src/components/MAPI/apps/AppList/utils.tsx

```diff
@@ -0,0 +1,327 @@
+        version.keywords.join(',').toLowerCase().includes(normalizedQuery):
+        return true;
+
+      default:
+        return false;
+    }
+  });
+}
+
+function makeAppPath(
+  name: string,
+  version: string,
+  catalogName: string
+): string {
+  return RoutePath.createUsablePath(AppsRoutes.AppDetail, {
+    app: name,
+    catalogName,
+    version,
+  });
+}
+
+function compareVersions(
+  a: IAppCatalogIndexAppVersion,
+  b: IAppCatalogIndexAppVersion
+) {
+  return compare(b.version, a.version);
+}
+
+interface IAppCatalogIndexResponse {
+  apiVersion: string;
+  entries: Record<string, IAppCatalogIndexAppVersion[]>;
+}
+
+/**
+ * Fetch the catalog index from the URL saved in the catalog CR.
+ * @param fetchFunc
+ * @param _auth
+ * @param catalog
+ */
+async function fetchAppCatalogIndex(
+  fetchFunc: (input: RequestInfo, init?: RequestInit) => Promise<Response>,
+  _auth: IOAuth2Provider,
+  catalog: applicationv1alpha1.IAppCatalog
+): Promise<IAppCatalogIndex> {
+  const url = normalizeAppCatalogIndexURL(catalog.spec.storage.URL);
+
+  // Enforce client-side CORS.
+  const response = await fetchFunc(url, { mode: 'cors' });
+  const responseText = await response.text();
+  const catalogIndexResponse = yaml.load(
+    responseText
+  ) as IAppCatalogIndexResponse;
+
+  const catalogIndex: IAppCatalogIndex = {
+    apiVersion: catalogIndexResponse.apiVersion,
+    entries: {},
+  };
+
+  for (const [appName, versions] of Object.entries(
+    catalogIndexResponse.entries
+  )) {
+    const sortedVersions = versions.sort(compareVersions);
+
+    const version = versions[0]?.version ?? 'n/a';
+    const iconURL = versions[0]?.icon ?? '';
+
+    catalogIndex.entries[appName] = {
+      catalogName: catalog.metadata.name,
+      catalogTitle: computeAppCatalogUITitle(catalog),
+      catalogIconUrl: catalog.spec.logoURL ?? '',
+      catalogIsManaged: isAppCatalogVisibleToUsers(catalog),
+      appIconURL: iconURL,
+      name: appName,
+      to: makeAppPath(appName, version, catalog.metadata.name),
+      versions: sortedVersions,
+    };
+  }
+
+  return catalogIndex;
+}
+
+export async function getAppCatalogsIndexAppList(
+  fetchFunc: (input: RequestInfo, init?: RequestInit) => Promise<Response>,
+  auth: IOAuth2Provider,
+  catalogs: applicationv1alpha1.IAppCatalog[]
+): Promise<IAppCatalogIndexAppList> {
+  const requests = catalogs.map((catalog) =>
+    fetchAppCatalogIndex(fetchFunc, auth, catalog)
+  );
+
+  const responses = await Promise.allSettled(requests);
+
+  const indexList: IAppCatalogIndexAppList = {
+    items: [],
+    errors: {},
+  };
+  for (let i = 0; i < responses.length; i++) {
+    const response = responses[i];
+
+    if (response.status === 'rejected') {
+      const catalogName = catalogs[i].metadata.name;
+      indexList.errors[catalogName] = extractErrorMessage(response.reason);
+      continue;
+    }
+
+    indexList.items.push(...Object.values(response.value.entries));
```

Comment:
Get error messages from requests that failed and get data from requests that didn't